### PR TITLE
Add warp-ctc implementation to Chainer model

### PR DESCRIPTION
**Pull request by jheymann85**
This PR adds warp-ctc to the Chainer model speeding up the training process (for me by roughly 35% using a GTX 980).

The new CTC method can be selected by specifying the new argument`ctctype`. There are two options:
`chainer`: Use the vanilla chainer implementation of CTC
`warpctc`: Use the warp-ctc wrapper

By default, all recipes use `chainer` for now. Maybe this can be changed in the future.

To find the warp-ctc shared library, the `path.sh` extends the `LD_LIBRARY_PATH` to include the warp-ctc build folder.

A small test verifying that the loss and the gradients are the same for both CTC types is included.

I also ran the WSJ recipe (without an external LM) and the results seem to be reasonable:

*eval92*: 18.4% WER
*dev93*: 24.2% WER

### CODE REVIEW

File: egs/chime4/asr1/run.sh [hunk 1]
```diff
@@ -171,7 +171,7 @@ if [ ${stage} -le 2 ]; then
 fi
 
 if [ -z ${tag} ]; then
-    expdir=exp/${train_set}_${etype}_e${elayers}_subsample${subsample}_unit${eunits}_proj${eprojs}_d${dlayers}_unit${dunits}_${atype}_aconvc${aconv_chans}_aconvf${aconv_filts}_mtlalpha${mtlalpha}_${opt}_bs${batchsize}_mli${maxlen_in}_mlo${maxlen_out}
+    expdir=exp/${train_set}_${etype}_e${elayers}_subsample${subsample}_unit${eunits}_proj${eprojs}_d${dlayers}_unit${dunits}_${atype}_aconvc${aconv_chans}_aconvf${aconv_filts}_mtlalpha${mtlalpha}_${opt}_bs${batchsize}_mli${maxlen_in}_mlo${maxlen_out}_ctc${ctctype}
```
[the remark this answers is not in the document]
Comment (jheymann85) [reply egs/chime4/asr1/run.sh]: Done

File: egs/chime4/asr1/run.sh [hunk 2]
```diff
@@ -216,7 +219,9 @@ if [ ${stage} -le 3 ]; then
         --maxlen-in ${maxlen_in} \
         --maxlen-out ${maxlen_out} \
         --opt ${opt} \
-        --epochs ${epochs}
+        --epochs ${epochs} \
+        --ctc_type ${ctctype} \
```
[the remark this answers is not in the document]
Comment (jheymann85) [reply egs/chime4/asr1/run.sh]: Done

File: tools/Makefile [hunk 3]
```diff
@@ -37,6 +37,17 @@ warp-ctc: venv/lib/python2.7/site-packages/torch
 	. venv/bin/activate; pip install cffi
 	. venv/bin/activate; cd warp-ctc/pytorch_binding && python setup.py install # maybe need to: apt-get install python-dev
 
+chainer_ctc: venv/bin/activate
+	git clone https://github.com/jheymann85/chainer_ctc.git
+	. venv/bin/activate; pip install cython
+	. venv/bin/activate; cd chainer_ctc && chmod +x install_warp-ctc.sh && ./install_warp-ctc.sh
+	. venv/bin/activate; cd chainer_ctc && pip install .
+	@echo --------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (jheymann85) [reply tools/Makefile]: Done

File: test/test_e2e_model.py [hunk 4]
```diff
@@ -72,10 +75,46 @@ def init_chainer_weight_const(m, val):
...
+            numpy.float32), tokenid=out_data)),
+        ("bbb", dict(feat=numpy.random.randn(100, 40).astype(
+            numpy.float32), tokenid=out_data)),
+        ("cc", dict(feat=numpy.random.randn(100, 40).astype(
+            numpy.float32), tokenid=out_data))
+    ]
+
+    def _propagate(ctc_type):
+        args = make_arg(
+            ctc_type=ctc_type, elayers=2, eunits=20, eprojs=20
```
Comment (ShigekiKarita) [inline test/test_e2e_model.py]: if you think these args `elayers=2, eunits=20, eprojs=20` are OK , make it default in `make_arg`

File: egs/wsj/asr1/run.sh [hunk 5]
```diff
@@ -25,6 +26,8 @@ elayers=6
 eunits=320
 eprojs=320
 subsample=1_2_2_1_1 # skip every n frame from input to nth layers
+# loss related
+ctctype=chainer
```
Comment (ShigekiKarita) [inline egs/wsj/asr1/run.sh]: @sw005320 do you think that `warpctc` should be default?

File: src/bin/asr_train.py [hunk 6]
```diff
@@ -291,6 +291,10 @@ def main():
     parser.add_argument('--subsample', default=1, type=str,
                         help='Subsample input frames x_y_z means subsample every x frame at 1st layer, '
                              'every y frame at 2nd layer etc.')
+    # loss
+    parser.add_argument('--ctc_type', default='chainer', type=str,
```
Comment (ShigekiKarita) [inline src/bin/asr_train.py]: @sw005320  ditto

File: egs/wsj/asr1/run.sh (hunk 5 above)
Comment (sw005320) [reply egs/wsj/asr1/run.sh]: I'll decide after I tested it with the other recipes. For now, it's OK to make chainer one as default.

File: test/test_e2e_model.py (hunk 4 above)
Comment (sw005320) [reply test/test_e2e_model.py]: @jheymann85 @jheymann85, if this is fixed (or if @jheymann85 think that it should be as it is, I'll merge this PR.

Comment (jheymann85) [reply test/test_e2e_model.py]: Ok, I kept the network configuration and adjusted the tolerance for the loss instead
(its `x: array(36.5968, dtype=float32)` vs. `y: array(36.596798, dtype=float32)`)

**Pull request merged**